clients/go/admin/pkce/auth_flow_orchestrator.go

```diff
@@ -0,0 +1,166 @@
+package pkce
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"github.com/pkg/browser"
+	"golang.org/x/oauth2"
+)
+
+const (
+	Timeout     = 15 * time.Second
+	RefreshTime = 5 * time.Minute
+)
+
+// TokenOrchestrator implements the main logic to initiate Pkce flow to issue access token and refresh token as well as
+// refreshing the access token if a refresh token is present.
+type TokenOrchestrator struct {
+	clientConfig *oauth2.Config
+	tokenCache   TokenCache
+}
+
+// RefreshToken attempts to refresh the access token if a refresh token is provided.
+func (f TokenOrchestrator) RefreshToken(ctx context.Context, token *oauth2.Token) (*oauth2.Token, error) {
+	ts := f.clientConfig.TokenSource(ctx, token)
+	var refreshedToken *oauth2.Token
+	var err error
+	refreshedToken, err = ts.Token()
+	if err != nil {
+		logger.Warnf(ctx, "failed to refresh the token due to %v and will be doing re-auth", err)
+		return nil, err
+	}
+
+	if refreshedToken != nil {
```

Comment:
can err != nil but refreshToken ever == nil ?

Reply:
yes.. refreshToken is optional..